nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -2624,8 +2624,31 @@ impl ArrayPrototype {
         todo!();
     }
 
-    fn to_reversed(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!();
+    fn to_reversed(agent: &mut Agent, this_value: Value, _: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? ToObject(this value).
+        let o = to_object(agent, this_value)?;
```

Comment:
> I tried to add a fast path similar to Array.prototype.reverse method but I wasn't able to figure out how to clone the array in order to not mutate the original array order. At least by doing array.clone() didn't work (it mutated the original array as well) and then the converting it to slice and back to an array triggers some issues between &[Option<Value>] and &[Value] that I couldn't find an easy fix - if we should add the fast path anyway, ideas are welcomed on how to implement it 🙏

As you noticed, we do not have an existing cloning algorithm for Arrays or their elements data. `array.clone()` just clones the Array reference, so this is the same as doing this in JS:

```ts
let array = [...];
let array2 = array; // array.clone();
```

In order to actually implement cloning, I spent a good while writing complicated and interesting ideas. Turns out Rust already has exactly what we need. If the Array is dense and trivial (contains no holes and no getters) then we can just use `Vec::extend_from_within(index..index+1)` to push an internal copy to the end of the Vec.

Where we need to use this the `ElementArrays` in `element_array.rs`: We can extend_from_within the correct `ElementArray2PowX` Vector with the index of our `from` Array's `elements: SealableElementsVector`. Now the last index in `ElementArray2PowX` is a copy of our `from` Array's elements data, and we can take that index as our clone'd Array's `elements: SealableElementsVector`. We also need to manually create an `ArrayHeapData` with that elements etc. but that's just side rumble at this point.

Once the cloning function is done, the last stop is to just do the `array.as_mut_slice(agent).reverse()` on the cloned Array.

Does this sound doable? Do you want to give it a go?

Reply:
Sure! I'll give it a try - I'm still not 100% clear on the implementation but let me revisit this along with the code and if I get questions I'll let you know :)

Reply:
I tried following your suggestion on Discord and committed my attempt for implementation 😃